### PR TITLE
[jindisk] Improve block cryption, reclaim and discard

### PR DESCRIPTION
**Pull request by lucassong-mh**
This PR includes:
1. Add symm cryption API (AES-128-CTR) and apply to `DiskArray` @cqs21 https://github.com/occlum/occlum/pull/1289
2. Refactor cryption API (refer to [rust-openssl](https://docs.rs/openssl/0.10.54/openssl/symm/index.html))
3. Improve block reclamation part in compaction
4. Rename `trim()` to `discard()` for more accuracy in block device layer. Add an implementation, leave performance impact `TODO`.

### CODE REVIEW

File: src/libos/crates/jindisk/src/util/disk_array.rs [hunk 1]
```diff
@@ -79,9 +84,9 @@ impl<T: Serialize> DiskArray<T> {
     }
 
     pub async fn persist(&self, _root_key: &Key) -> Result<()> {
-        // TODO: Add encryption(symmetric) for blocks
         for (hba, block) in self.table.iter() {
-            self.disk.write(*hba, block.as_slice()).await?;
+            let ciphertext = DefaultCryptor::symm_encrypt_block(block.as_slice(), &self.key)?;
```
Comment (tatetian) [inline src/libos/crates/jindisk/src/util/disk_array.rs]: Is there security weakness in using the same key for every block? I thought the crypto API should take an argument for user to decide the IV. This way, different block can be encrypted with different IV, although the key is the same. Should our crypto API allow providing IVs?

File: src/libos/crates/jindisk/src/util/cryption.rs [hunk 2]
```diff
@@ -25,28 +25,32 @@ pub type Iv = [u8; AUTH_ENC_IV_SIZE];
...
 
-    /// Encrypt a block using symmetric key cipher(AES128 CTR mode)
-    fn symm_encrypt_block(data: &[u8], key: &Key) -> Result<[u8; BLOCK_SIZE]>;
+    /// Encrypt a block. (Non-AEAD)
+    fn encrypt_block(data: &[u8], key: &Key) -> Result<[u8; BLOCK_SIZE]>;
 
-    /// Decrypt a block using symmetric key cipher(AES128 CTR mode)
-    fn symm_decrypt_block(data: &[u8], key: &Key) -> Result<[u8; BLOCK_SIZE]>;
+    /// Decrypt a block. (Non-AEAD)
+    fn decrypt_block(data: &[u8], key: &Key) -> Result<[u8; BLOCK_SIZE]>;
```
Comment (tatetian) [inline src/libos/crates/jindisk/src/util/cryption.rs:53]: The current crypto API encourages copying by returning `[u8; BLOCK_SIZE]`. Usually, the user of such API can prepare the target enc/dec buffer before calling the API. So by accepting `&mut [u8]` as input, the API can save the user from one extra copy of a block of memory.

**Review by tatetian — COMMENTED** [review]
Is there any performance degradation introduced by this PR?

The PR looks mostly good to me. I raised some minor issues.

File: src/libos/crates/jindisk/src/util/cryption.rs (hunk 2 above)
Comment (lucassong-mh) [reply src/libos/crates/jindisk/src/util/cryption.rs]: The start point to design this interface is to ensure the invariant of block-granularity cryption. There is a non-copy version called `encrypt_arbitrary(plaintext, ciphertext)` can serve the purpose.

And no obvious performance degradation according to some microbenchmarks by this PR.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**